onnxruntime/core/providers/cpu/math/cumsum.cc

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "cumsum.h"
+#include "core/providers/cpu/tensor/utils.h"
+#include "core/framework/op_kernel.h"
+#include "core/framework/tensorprotoutils.h"
+
+using namespace onnxruntime;
+
+namespace {
+// static section
+std::vector<int64_t> GetStarts(int64_t rank, int64_t axis, int64_t index) {
+  std::vector<int64_t> starts(rank, 0);
+  starts[axis] = index;
+  return starts;
+}
+template <typename T>
+void ZeroOutSliceAtIndex(Tensor& output, int64_t rank, int64_t axis, int64_t index,
+                         const std::vector<int64_t>& slice_dims, const std::vector<int64_t>& steps, const int64_t slice_size) {
+  T zero{};
+  auto output_starts(GetStarts(rank, axis, index));
+  WritableSliceIterator<T> output_iterator(output, output_starts, slice_dims, steps);
+  for (int64_t k = 0; k < slice_size; ++k, ++output_iterator) {
+    *output_iterator = zero;
+  }
+}
+template <typename T>
+void CopySlices(const Tensor& input, Tensor& output,
+                std::vector<int64_t>& input_starts, const std::vector<int64_t>& output_starts,
+                const std::vector<int64_t>& slice_dims, const std::vector<int64_t>& steps, const int64_t slice_size) {
+  SliceIterator<T> input_iterator(input, input_starts, slice_dims, steps);
+  WritableSliceIterator<T> output_iterator(output, output_starts, slice_dims, steps);
+  for (int64_t k = 0; k < slice_size; ++k, ++output_iterator, ++input_iterator) {
+    *output_iterator = *input_iterator;
+  }
+}
+template <typename T>
+void SumSlices(const Tensor& input, Tensor& output,
+               std::vector<int64_t>& input_starts, const std::vector<int64_t>& output_starts, const std::vector<int64_t>& previous_output_starts,
```

Comment:
can input_starts be passed by const-ref?

Reply:
> can input_starts be passed by const-ref?

Modified to const by ref.